### ISSUE TITLE
VidHD: support 80-column video mode for II/II+

### ISSUE THREAD

**Issue by tomcw** [AppleWin/AppleWin#1323]
(Spun off from #1321)

AppleWin 1.30.19 (or earlier) currently provides only partial VidHD support.
See the comments here for some info: https://github.com/AppleWin/AppleWin/blob/1127447a10b0cdbf0ec921311dfb91b7554351fe/source/VidHD.cpp#L24-L34

There's also only limited support for the card's firmware:
- Only the first 3 bytes at $C300-$C302 are returned, the remaining addresses returning floating bus values.

Essentially this is _just enough_ to allow all Apple II models to support the IIgs' Super Hi-Res (SHR) video modes, eg. as supported by Total Replay.

This issue is to:
- understand why ProDOS sets MACHID bit1 (ie. 80 column card present) for a II/II+ with a VidHD in slot 3 for AppleWin 1.30.19 (or earlier)
- decide whether to include the full VidHD firmware (eg. for complete card ROM ID detection & to support `JSR C300` / `PR#3`)
- consider whether to support 80-column video for II/II+

ref. VidHD manual 1.2 [here](https://www.callapple.org/docs/vidhd/VidHD_Manual_1.2.pdf)

**Comment by tomcw** [comment]
Test Apple II+ with VidHD in slot 3:

1) Boot ProDOS 2.4.3, and read MACHID ($BF98) - which returns $62:
![DSC_1616](https://github.com/user-attachments/assets/decd54cd-f2d5-4835-953f-a024a2454b40)

2a) Boot, but without any ProDOS/DOS. Then activate VidHD firmware (`PR#3`) and issue `<ESC> 8` to switch to 80x24 resolution (see VidHD manual), showing monitor output: (40-col)
![DSC_1615](https://github.com/user-attachments/assets/97df3dc0-2168-4cba-a389-d61c934086d8)

2b) ...and the corresponding HDMI output: (ie. 80-col)
![DSC_1614](https://github.com/user-attachments/assets/bb18f6cd-db23-484a-b468-b2cf0e8898a9)

**Comment by tomcw** [comment]
I've taken a look at this:
>understand why ProDOS sets MACHID bit1 (ie. 80 column card present) for a II/II+ with a VidHD in slot 3 for AppleWin 1.30.19 (or earlier)

Using AppleWin 1.30.19, setup as an Apple II+ with s3=VidHD (and s5=HDC, s6=DiskII);
- ProDOS attempts to ID the card using $Csxx ID bytes:
   - `$28be: ID bytes: -- 20 -- 00 -- 03` (Block device: Disk II, HDC)
   - `$2785: ID bytes: 08 -- 28 -- 58 -- 70`
- Else at $27d3 determine if there is ROM at $Cs00
   - Check this single $Cs00 byte is not $FF and then reads the same value 256 times
   - If this succeeds then $bf99 is a bitmap of cards with ROMs where bit offset n -> slot n (eg. bit3 / $08 is for slot 3)
 - $21fc: special function for slot 3 card and when $bf99 & 8 is true (ie. $C3nn ROM detected):
    - this uses //e+ software switches even for the II+!
    - it uses 80STORE ON and then PAGE2 to check values written/read at $478
    - on a //e without 80-col card in s3: this will return floating bus for the aux mem (I assume)
    - but on a II+ (where these soft switches don't exist) this just  writes/read main mem, and so passes this check, indicating that slot 3 is an 80 column card.

**Comment by tomcw** [comment]
Here's my _rough_ disassembly (and notes) of what I traced through for the above analysis:
[GH1323-VidHD-BF98.txt](https://github.com/user-attachments/files/17035153/GH1323-VidHD-BF98.txt)

**Comment by tomcw** [comment]
NB. A Mockingboard's 6522 register $0 (at $Cs00) at power-on (and after a reset) will return $FF.
So the code at $27d3 to determine if there is ROM at $Cs00 will correctly fail for a Mockingboard.
Also $FF is an invalid 6502 opcode (and a NOP for 65C02).

**Comment by tomcw** [comment]
Also note that the VidHD $Csnn bytes (see 80-col photo above) don't match either of the card detection values.
But it will pass the check indicating it has a ROM at $Csnn.
And then (even in a II+) it will support the //e soft switches, so will pass the test for the 80-col card.

**Comment by xotmatrix** [comment]
Because of the ProDOS detection bug, it would pass even if it didn't respond to the soft switches. As long as there is a ROM in slot 3, it thinks there is an 80-column card.

**Comment by tomcw** [comment]
>Because of the ProDOS detection bug, it would pass even if it didn't respond to the soft switches.

Yes... for an Apple II/II+ without a VidHD in slot 3 (and also with a VidHD in slot 3, where the soft switches work).

For an Apple //e, it will work correctly.
